### PR TITLE
qemu/tests/: Add new case for block hotplug

### PR DESCRIPTION
**Pull request by spcui**
Signed-off-by: Shuping Cui <scui@redhat.com>

ID: 1081682 1081684

### CODE REVIEW

File: qemu/tests/block_hotplug.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    :param params: Dictionary with the test parameters.
+    :param env:    Dictionary with test environment.
+    """
+    def find_image(blk_num):
+        image_params = params.object_params("%s" % img_list[blk_num + 1])
+        o = storage.get_image_filename(image_params, data_dir.get_data_dir())
+        return o
+
+    def find_disk(session, cmd):
+        disks = []
```
Comment (xutian) [inline qemu/tests/block_hotplug.py]: Useless declare

File: qemu/tests/block_hotplug.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    1) Boot up guest with/without block device(s).
+    2) Hoplug block device and verify
+    3) Do read/write data on hotplug block.
+    4) Unplug block device and verify
+
+    :param test:   QEMU test object.
+    :param params: Dictionary with the test parameters.
+    :param env:    Dictionary with test environment.
+    """
+    def find_image(blk_num):
```
Comment (xutian) [inline qemu/tests/block_hotplug.py]: where your docstring???

File: qemu/tests/block_hotplug.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        else:
+            pattern = "^\d+"
+            cmd = params.get("get_disk_index", "wmic diskdrive get index")
+
+        output = session.cmd_output_safe(cmd)
+        disks = re.findall(pattern, output, re.M)
+        return disks
+
+    def get_new_disk(disk1, disk2):
+        disk = list(set(disk2).difference(set(disk1)))
```
Comment (xutian) [inline qemu/tests/block_hotplug.py]: do you really need to convert it to list?

File: qemu/tests/block_hotplug.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        return disk
+
+    img_list = params.get("images").split()
```
Comment (xutian) [inline qemu/tests/block_hotplug.py]: params.object_params("images")

File: qemu/tests/block_hotplug.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+
+    for i in xrange(repeat_times):
+        error_context.context("Hotplug block device (iteration %d)" % i,
+                              logging.info)
+
+        session = vm.wait_for_login(timeout=timeout)
+        vm.devices.set_dirty()
```
Comment (xutian) [inline qemu/tests/block_hotplug.py]: dirty flag should be set after hotplug/unplug, right??

File: qemu/tests/block_hotplug.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+            utils_test.run_virt_sub_test(test, params, env, sub_type)
+
+        for num in xrange(blk_num):
+            device = qdevices.QDevice(pci_type)
+            if params.get("need_plug") == "yes":
+                if params.get("need_controller", "no") == "yes":
+                    controller_model = params.get("controller_model")
+                    controller = qdevices.QDevice(controller_model)
+                    controller.hotplug(vm.monitor)
+                    controller.verify_hotplug("", vm.monitor)
```
Comment (xutian) [inline qemu/tests/block_hotplug.py]: view from code this line do nothing!!

File: qemu/tests/block_hotplug.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+                drive.set_param("format", img_format_type)
+                drive_id = drive.get_param("id")
+                drive.hotplug(vm.monitor)
+                time.sleep(pause)
+
+                device.set_param("drive", drive_id)
+                device.set_param("id", "block%d" % num)
+                device.hotplug(vm.monitor)
+                time.sleep(pause)
+                device.verify_hotplug("", vm.monitor)
```
Comment (xutian) [inline qemu/tests/block_hotplug.py]: same question

File: qemu/tests/block_hotplug.py (hunk 5 above)
Comment (spcui) [inline qemu/tests/block_hotplug.py]: yes

**Comment by spcui** [comment]
@xutian updated with your comments, could you review again?

**Comment by xutian** [comment]
Please update you commit message.

File: qemu/tests/block_hotplug.py (hunk 5 above)
Comment (xutian) [inline qemu/tests/block_hotplug.py]: so, please update you code

**Comment by Zhengtong** [comment]
On Power platform, the test could be passed with "virtio-scsi" which is set to be the system disk type. For the test with "virtio_blk(system disk)" and without plug virtio_blk type disk,  the case will erase the guest system disk data, as the system disk will be vdb and the tested disk will be vda, this a different behavior from X86.  And this is the problem we always talk about.

**Comment by spcui** [comment]
@Zhengtong you can ignore this case on power platform first, will support it in the further, thanks your comment.

**Comment by Zhengtong** [comment]
Acked by Zhengtong   <zhengtli@redhat.com>
Since the script passed on other variants. the scripts is ok for me.

**Comment by spcui** [comment]
@xutian updated, could you review again? thanks

File: qemu/tests/block_hotplug.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        """
+        image_params = params.object_params(image_name)
+        o = storage.get_image_filename(image_params, data_dir.get_data_dir())
+        return o
+
+    def find_disk(session, cmd):
+        """
+        Find all disks in guest.
+        """
+        if params.get("os_type") == "linux":
```
Comment (xutian) [inline qemu/tests/block_hotplug.py]: only find disk vd[a-z], so your test only support virtio-blk test ? how about virtio-scsi ?

File: qemu/tests/block_hotplug.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+        output = session.cmd_output_safe(cmd)
+        disks = re.findall(pattern, output, re.M)
+        return disks
+
+    def get_new_disk(disk1, disk2):
+        """
+        Get the different disk between disk1 and disk2.
+        """
+        disk = list(set(disk2).difference(set(disk1)))
```
Comment (xutian) [inline qemu/tests/block_hotplug.py]: why need to convert from set to list?

File: qemu/tests/block_hotplug.py (hunk 8 above)
Comment (spcui) [inline qemu/tests/block_hotplug.py]: for virtio-scsi, get_disk_pattern = "^/dev/sd[a-z]*$"  set it in cfg file.

File: qemu/tests/block_hotplug.py (hunk 9 above)
Comment (spcui) [inline qemu/tests/block_hotplug.py]: need to return a list, so convert it.

**Comment by xutian** [comment]
@Zhengtong  according your comments this case can always pass in power platform, right?

**Comment by Zhengtong** [comment]
@xutian No,  only can't pass with "virtio-blk(system disk) + without_hotplug + virtio-blk(data disk)".

**Comment by xutian** [comment]
@Zhengtong  got it.
@spcui  please add no powerfilter on @Zhengtong mentioned scene

**Comment by spcui** [comment]
@xutian updated, could you review again? thanks

File: generic/tests/cfg/block_hotplug.cfg [hunk 10]
```diff
@@ -2,47 +2,84 @@
...
-            image_format_stg = raw
-            image_format_stg0 = raw
-            image_format_stg1 = raw
+
     variants:
         - one_pci:
-            pci_num = 1
-            repeat_times = 300
+            blk_num = 1
+            repeat_times = 10
```
Comment (xutian) [inline generic/tests/cfg/block_hotplug.cfg]: why decrease repeat times?

File: qemu/tests/block_hotplug.py [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+                        raise exceptions.TestFail(err)
+
+                disks_before_plug = find_disk(session, get_disk_cmd)
+
+                drive = qdevices.QRHDrive("block%d" % num)
+                drive.set_param("file", find_image(img_list[num + 1]))
+                drive.set_param("format", img_format_type)
+                drive_id = drive.get_param("id")
+                drive.hotplug(vm.monitor)
+                time.sleep(pause)
```
Comment (xutian) [inline qemu/tests/block_hotplug.py]: do we need sleep here, I think sleep should be a line#131

File: qemu/tests/block_hotplug.py (hunk 9 above)
Comment (xutian) [inline qemu/tests/block_hotplug.py]: ok, see it

File: qemu/tests/block_hotplug.py (hunk 8 above)
Comment (xutian) [inline qemu/tests/block_hotplug.py]: ok, see it.

File: qemu/tests/block_hotplug.py (hunk 11 above)
Comment (spcui) [inline qemu/tests/block_hotplug.py]: I think need sleep after hot-plugging the drive and device.

Comment (xutian) [inline qemu/tests/block_hotplug.py]: do you really have a test about it.

Comment (spcui) [inline qemu/tests/block_hotplug.py]: yes

Comment (xutian) [inline qemu/tests/block_hotplug.py]: Please test remove  sleep in line#103 and line#108 , then add sleep in line#131.

**Comment by spcui** [comment]
PASS smp_16.16384m.repeat1.run_test.Host_RHEL.m7.u2.qcow2.virtio_blk.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.multi_pci
PASS smp_16.16384m.repeat1.run_test.Host_RHEL.m7.u2.qcow2.virtio_blk.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.multi_pci

**Comment by xutian** [comment]
fine. thanks. Merged

**Pull request merged**